[영화 티켓 예매 3, 4단계] 오이 미션 제출합니다.

cucumber99 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
- MVP 패턴을 처음 접하였는데, 일반적으로 MVP 패턴을 나타내는 그림을 보면 `View`는 오직 `Presenter`에게만 의존하고 있어요. 근데 제 코드에서는 `View`에서 `Intent`를 통해 특정 도메인 객체를 얻어오고, `Presenter`에게 도메인 객체를 인자로 받는 경우도 있기 때문에 의존성을 제거할 수 없었어요. 만약 제거해야 한다면, UI Model같은 개념을 도입해야 하나요?
- `Presenter` 테스트에서 `Presenter`의 어떤 메소드를 호출했을 때 `View`의 특정 메소드들이 호출되었는지 확인하는 방식으로만 테스트를 진행했는데, 이것이 올바른 테스트 방식인가요? 아직까지 통합 테스트의 개념에 대해 잘 이해하지 못한 것 같아요🥲
- 추가로, `RecyclerView`를 적용하고 테스트를 하기 위해서 `espresso-contrib`에 대한 의존성을 추가하였는데, 의존성 충돌이 일어나서 테스트를 진행하지 못해 우선 삭제했습니다.. 

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

https://github.com/user-attachments/assets/fd733438-c1eb-443a-875d-8783eae5f499

cucumber99 committed:
chore: 디렉터리 구조 변경

cucumber99 committed:
feat: 상영 중인 영화목록 객체 구현

cucumber99 committed:
feat: MoviesContract 추가

cucumber99 committed:
feat: MoviesPresenter 구현

cucumber99 committed:
feat: MoviesActivity MVP 패턴 적용

cucumber99 committed:
chore: MockK 의존성 추가

cucumber99 committed:
test: MoviesPresenter 테스트

- 영화 목록 출력을 테스트합니다
- 버튼 클릭을 테스트합니다

cucumber99 committed:
feat: BookingContract 작성

cucumber99 committed:
feat: MoviePresenter 구현

cucumber99 committed:
feat: BookingActivity MVP 패턴 적용

cucumber99 committed:
test: BookingPresenter 테스트

cucumber99 committed:
feat: BookingSummaryContract 작성

cucumber99 committed:
feat: BookingSummaryPresenter 구현

cucumber99 committed:
feat: BookingSummaryActivity MVC 패턴 적용

cucumber99 committed:
test: BookingSummaryPresenter 테스트

cucumber99 committed:
chore: 도메인 패키지 변경

cucumber99 committed:
docs: 기능 구현 사항 추가

cucumber99 committed:
chore: 요금 정책 제거

cucumber99 committed:
feat: 문자열 리소스 및 색상 추가

cucumber99 committed:
feat: 좌석 등급 열거형 클래스 구현

cucumber99 committed:
feat: 위치 클래스 구현

cucumber99 committed:
feat: 좌석 클래스 구현

cucumber99 committed:
refactor: 네이밍 및 코드 개선

cucumber99 committed:
test: 좌석 등급 테스트

cucumber99 committed:
test: 좌석 정보 테스트

cucumber99 committed:
feat: 선택 좌석 클래스 구현

cucumber99 committed:
test: 선택 좌석 테스트

cucumber99 committed:
feat: SeatContract 작성

cucumber99 committed:
feat: SeatPresenter 구현

cucumber99 committed:
feat: 좌석 선택 UI 구현

cucumber99 committed:
feat: 좌석 선택 액티비티 구현

cucumber99 committed:
test: SeatsActivity 테스트

cucumber99 committed:
feat: 예매 시점과 선택한 상영 시간 비교 로직 추가

cucumber99 committed:
refactor: 기타 리팩터링

cucumber99 committed:
test: SeatsPresenter 테스트

cucumber99 committed:
test: 테스트명 변경

cucumber99 committed:
feat: SeatPosition 직렬화 추가

cucumber99 committed:
feat: 좌석 출력 기능 구현

cucumber99 committed:
refactor: 일부 UI 수정

cucumber99 committed:
feat: 상영 영화 정보 및 포스터 추가

cucumber99 committed:
feat: 광고 레이아웃 및 사진 추가

cucumber99 committed:
feat: 출력 항목 인터페이스 및 구현체 추가

cucumber99 committed:
refactor: ListView -> RecyclerView 대체

cucumber99 committed:
refactor: ArrayAdapter -> ListAdapter 대체 및 광고 출력 대응

cucumber99 committed:
chore: MoviesActivityTest 삭제

- 의존성 문제로 인해 테스트 코드를 삭제합니다

cucumber99 committed:
refactor: 좌석 경고 문구 수정

cucumber99 committed:
feat: Configuration Change 대응

cucumber99 committed:
refactor: ktlint 적용

cucumber99 committed:
refactor: 불필요한 상수 제거

cucumber99 committed:
test: 날짜 오류 수정

cucumber99 committed:
test: 오류 메시지 출력 테스트 수정

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/presentation/movies/MoviesActivity.kt:
```diff
@@ -0,0 +1,62 @@
+            }
+        }
+        return result
+    }
```
```
영화 목록에 영화가 세 번 노출될 때마다 광고가 한 번 노출된다.
```

영화 목록에서 3개마다 1개씩 광고가 들어가는 것은 어떻게 보면 비즈니스적인 로직이라고 볼 수도 있지 않을까 싶네요. 지금은 광고를 로컬에서 갖고 오고 있지만 실제로 광고 데이터는 어떠한 서버 혹은 SDK로 부터 받아서 사용을 하게 될텐데요. 그런 상황이라고 가정을 하고 해보시면 조금 더 쉽게 접근이 될 것 같네요

cucumber99 replied:
별도의 도메인 객체가 광고를 삽입하게끔 변경하였습니다!

[feat: 광고 추가 클래스 구현](https://github.com/woowacourse/android-movie-ticket/pull/142/commits/497c19bc3c2603820cf98b022e418781027e1a34)
[refactor: 광고 기능 분리](https://github.com/woowacourse/android-movie-ticket/pull/142/commits/9a2412b91080fcf0044e4aa56ddbbdbc74a0d2af)

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/presentation/booking/BookingActivity.kt:
```diff
@@ -0,0 +1,183 @@
+import java.time.LocalTime
+import java.util.Locale
+
+class BookingActivity : BaseActivity(), BookingContract.View {
+    override val layoutRes: Int
+        get() = R.layout.activity_booking
+
+    private lateinit var presenter: BookingPresenter
+    private lateinit var movie: Movie
+    private var dateItemPosition: Int = DEFAULT_POSITION
+    private var timeItemPosition: Int = DEFAULT_POSITION
```
MVP 관점에서 봤을 때 Activity는 무엇인가요? View인가요? Presenter인가요?

이 질문에 대한 다변에 따라서 지금 전역으로 선언한 몇몇 변수들이 어디에 위치해야 할지도 결정이 나게 될겁니다.

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/domain/model/seat/SelectedSeats.kt:
```diff
@@ -0,0 +1,33 @@
+            _seats.remove(seat)
+            return
+        } else {
+            require(_seats.size < headCount) {
+                String.format(Locale.getDefault(), SELECT_MESSAGE, headCount)
+            }
```
아무리 봐도 View에 표시해야 할 데이터를 model에서 결정해 주는 것으로 보이는데요.

레벨 1로 돌아가서 이 코드는 어떠한 문제가 있을까요?

cucumber99 replied:
도메인이 화면 출력 형식을 결정짓는다는 문제가 있습니다.
도메인에서는 오류 메시지만 반환하고, 형식은 외부에서 결정짓도록 수정했습니다.
```kotlin
val message = e.message?.let {
    String.format(Locale.getDefault(), it, movieTicket.headCount)
} ?: e.stackTraceToString()
view.showToast(message)
```

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/presentation/booking/BookingPresenter.kt:
```diff
@@ -0,0 +1,66 @@
+    private var headCount: HeadCount = HeadCount()
+    private val movieScheduler: MovieScheduler by lazy { MovieScheduler(movie.startScreeningDate, movie.endScreeningDate) }
+
+    override fun onViewCreated() {
+        view.initBooking()
+        view.showMovie(movie)
+        view.showBookableDates(movieScheduler.getBookableDates())
+        view.updateHeadCount(headCount.value)
+    }
```
`onViewCreated`, `onConfigurationChanged`  같은 메서드 이름이나 역할이 Android View 생명주기에 너무 강하게 의존되어 있는 것 같아 보여요.

이로 인해 `BookingContract.Presenter`는 특정 플랫폼의 생명주기 변화에 직접 반응해야 해서, 플랫폼 독립적인 비즈니스 로직과 분리가 어려워질 수 있어요.

```
"만약 View가 Android가 아니거나, 다른 생명주기를 가진 환경이라면 어떻게 될까?"
```
를 한 번 고민해 보면 좋을 것 같아요.

이런 상황에서는 `프레젠터는 플랫폼에 독립적이어야 한다`는 관점으로 접근할 수도 있으니 한 번 천천히 고민해 봅시다

cucumber99 replied:
메소드 네이밍이 안드로이드 생명주기에 의존하지 않도록 변경하였습니다!

[refactor: 좌석 선택 화면 및 프레젠터 개선](https://github.com/woowacourse/android-movie-ticket/pull/142/commits/1c4d1fa00f1dbc4337e2500a6c8773c3854b0d8c)

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/presentation/movies/MoviesActivity.kt:
```diff
@@ -0,0 +1,62 @@
+        moviesView.adapter = adapter
+    }
+
+    override fun navigateToBooking(movie: Movie) {
+        val intent =
+            Intent(this, BookingActivity::class.java).apply {
+                putExtra(IntentKeys.MOVIE, movie)
+            }
+        startActivity(intent)
+    }
```
> MVP 패턴을 처음 접하였는데, 일반적으로 MVP 패턴을 나타내는 그림을 보면 View는 오직 Presenter에게만 의존하고 있어요. 근데 제 코드에서는 View에서 Intent를 통해 특정 도메인 객체를 얻어오고, Presenter에게 도메인 객체를 인자로 받는 경우도 있기 때문에 의존성을 제거할 수 없었어요. 만약 제거해야 한다면, UI Model같은 개념을 도입해야 하나요?

이런 상황에서 제가 생각하는 가장 베스트는 UI Model을 만드는 것이긴 합니다.

그런데 상황에 따라서 다른 방향도 생각할 수가 있는데요. `이 Model만의 유일한 정보 값안 넘긴다.` 라고 생각을 하면 그 값을 갖고 `어딘가에 요청을 해서 Movie` 객체를 받아오는 형태도 가능하게 됩니다. 그런데 지금 상황에서는 가장 쉬운 방법은 `UI Model`로 보이긴 하지만 다른 시각의 방법도 한 번 고민해보시기 바랍니다.

cucumber99 replied:
> 이 Model만의 유일한 정보 값안 넘긴다. 라고 생각을 하면 그 값을 갖고 어딘가에 요청을 해서 Movie 객체를 받아오는 형태도 가능하게 됩니다. 

Intent을 통해 도메인 객체를 직접 넘기지 않고 도메인 객체가 가진 원시값만 넘긴 다음,
해당 원시값을 통해 그 값을 가지고 있는 객체를 받아오는 방식으로 이해했는데요.
그렇다면 도메인 객체들을 저장하는 역할을 담당하는 별도의 계층이 존재해야 한다는 뜻인가요?

lee-ji-hoon replied:
말씀 주신 `별도의 계층`이라는 선택지가 필요로 하게 되는데, 이 방법은 지금은 너무 오버 스펙이니 그냥 인지만 하고 계셔도 충분할 것 같네요.

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/presentation/seats/SeatsActivity.kt:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    companion object {
+        private const val TICKET_INTENT_ERROR = "[ERROR] 예매 정보에 대한 키 값이 올바르지 않습니다."
```
`strings.xml` 👀

cucumber99 replied:
헉.. 예외 메시지를 모두 분리했습니다!

[refactor: 오류 메시지 분리](https://github.com/woowacourse/android-movie-ticket/pull/142/commits/101aa0f22f6924582572e80cad962a5925983c14)

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/presentation/seats/SeatsContract.kt:
```diff
@@ -0,0 +1,43 @@
+            y: Int,
+        ): Seat
+
+        fun getSelectedSeats(): List<Seat>
+
+        fun isSelectedSeat(seat: Seat): Boolean
```
MVP 관점에서 `~get` `~is`처럼 어떠한 값을 반환하는 게 맞는 패턴인지 고민해 보시기 바랍니다 🤔

cucumber99 replied:
MVP 패턴에 대한 이해도가 부족해서 Presenter가 데이터를 반환하는 형태를 갖춘 것 같아요.
근데 `getSelectedSeats`의 경우에는 View에서 Configuration change가 일어날 때 현재 좌석들의 상태를 복원하기 위해
Presenter가 좌석들의 정보를 반환하게끔 하여 `Bundle`에 `List<Seat>`을 저장하고 있었습니다.

Presenter 내부의 도메인 객체가 가진 데이터를 반환하지 않고 상태를 저장하는 방법을 찾아봤는데, 
데이터를 캐싱하는 싱글톤 객체를 선언하여 사용하는 방법을 추천하고 있어요. [관련 문서](https://medium.com/@theMikhail/presenters-are-not-for-persisting-f537a2cc7962#.ssl022wg7)
물론 위 문서에서는 네트워크에서 얻어온 데이터를 출력할 때, Configuration change가 일어나면 데이터를 재요청하지 않기 위해 위와 같은 방법을 채택한 것 같아서 제 문제와는 다른 것 같습니다.

코니는 이 방법에 대해 어떻게 생각하시나요?

lee-ji-hoon replied:
말해주신 대로 오이의 문제와 다른 것입니다. 제가 이야기하고자 했던 내용은 `View` 가 원할 때 어떠한 `Model` 을 `Presenter`에게 요청을 해서 반환하는 구조가 `MVP` 에서 맞는지에 대한 질문이었습니다.

`getSelectedSeats` 의 경우 `Configuration change` 의 경우 현재 좌석들의 상태를 굳이 `return`을 해서 보여줘야 했을까요? 

1. View는 Presenter에게 현재 화면을 복구해야 한다고 알린다.
2. Presenter는 화면을 복구하기 위한 Model을 만든다. 
3. Presenter는 View에게 해당 Model로 Update 할 것을 알린다.

`MVP` 라면 이런 구조로 가야 했는데, `상태를 저장하는 방법`이 고민이 됐던 것은 오이랑 나눈 DM에서 해결이 됐길 바랍니다. 혹시 해결이 안 됐다면 편하게 연락 주세요

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/ui/adapter/MovieAdapter.kt:
```diff
@@ -4,48 +4,57 @@ import android.content.Context
-    override fun getView(
-        position: Int,
-        convertView: View?,
+class MovieAdapter(private val onClick: (Movie) -> Unit) : ListAdapter<MoviesItem, RecyclerView.ViewHolder>(diffCallback) {
```
일반 RecyclerView Adapter와 ListAdapter는 어떠한 차이점이 있을까요?

cucumber99 replied:
`ListAdapter`는 `RecyclerView.Adapter`를 확장한 것으로, 데이터 변경 사항을 자동으로 반영해줍니다.
`submitList`로 출력할 항목들을 전달한 다음, `DiffUtil.ItemCallback`을 구현하여 간편하게 처리할 수 있습니다.

lee-ji-hoon commented on app/src/main/res/layout/activity_seats.xml:
```diff
@@ -0,0 +1,387 @@
+
+    </TableLayout>
+
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:layout_width="match_parent"
+        android:layout_height="80dp"
+        android:background="@color/white"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintBottom_toBottomOf="parent">
```
이미 최상위에 `ConstraintLayout` 이기에 여기 `ConstraintLayout` 은 없이도 구현이 되지 않았을까요? XML View를 다루다 보면 계층에 대해서 되게 민감하게 반응을 해야 하는데 그 이유는 Slack에서 단체로 질문을 드렸던 내용에서 충분히 다뤘으니 한 번 도전해 보시면 좋겠습니다.

> 근데 다음 미션도 있으니 인지만 하고 넘어가셔도 됩니다.

lee-ji-hoon commented on app/src/test/java/woowacourse/movie/presentation/booking/BookingPresenterTest.kt:
```diff
@@ -0,0 +1,104 @@
+
+    @BeforeEach
+    fun setUp() {
+        view = mockk(relaxed = true)
```
`relaxed = true` 는 어떠한 역할을 하나요?

cucumber99 replied:
메소드에 전달되는 값을 정의해주지 않아도 테스트할 수 있도록 합니다!
`LocalDateTime.now()`에 의존적인 기능을 테스트하기 위해서 사용했습니다.

lee-ji-hoon commented on app/src/test/java/woowacourse/movie/presentation/booking/BookingPresenterTest.kt:
```diff
@@ -0,0 +1,104 @@
+import java.time.LocalDateTime
+import java.time.LocalTime
+
+class BookingPresenterTest {
```
> Presenter 테스트에서 Presenter의 어떤 메서드를 호출했을 때 View의 특정 메서드들이 호출되었는지 확인하는 방식으로만 테스트를 진행했는데, 이것이 올바른 테스트 방식인가요? 아직까지 통합 테스트의 개념에 대해 잘 이해하지 못한 것 같아요🥲

우선 통합 테스트를 Presenter로 하는 것은 부적합합니다. 통합 테스트의 경우 

```
Presenter + View + 실제 데이터 흐름까지 포함해서, 진짜로 UI가 뜨고 상호작용하는 걸 테스트
```

하는 것을 메인으로 봐야 하는데요. Android에서는 `ActivitySenario` 로 하게 됩니다. 지금은 단위 테스트이기에 `view` 의 경우 `verify`를 사용해서 `실제로 호출이 됐는지` 정도만 확인을 하게 되는 것이 보편적이기에 지금 테스트가 충분해 보입니다.

lee-ji-hoon reviewed (changes requested):
오이 안녕하세요! 우선 마지막 단계까지 고생 많았습니다.

남겨주신 질문에 대해서 제 의견을 남겨뒀으며, 필수적으로 확인해봤으면 하는 곳들에 리뷰를 남겨놨습니다.

앞으로 남은 미션들도 있기에 오이가 되는 만큼 리뷰를 반영하고 다시 리뷰 요청 주시기 바랍니다!

cucumber99 committed:
refactor: 좌석 선택 화면 및 프레젠터 개선

- 프레젠터 메소드 네이밍이 안드로이드 생명주기에 의존하지 않습니다
- 프레젠터가 특정 값을 반환하지 않습니다
- 태그에 SeatPosition 객체를 저장합니다

cucumber99 committed:
refactor: 예외 메시지 변경

cucumber99 committed:
refactor: 오류 메시지 분리

cucumber99 committed:
feat: 광고 추가 클래스 구현

cucumber99 committed:
refactor: 광고 기능 분리

lee-ji-hoon reviewed (approved):
오이 안녕하세요 영화 티켓 예매 미션 고생 많았습니다.

DM으로 질문을 주셨던 부분과 코멘트로 남겨주신 내용들에 대해서 답변을 달아뒀으며, 이해가 안 되거나 어려운 부분이 있다면 추가로 연락해 주시기 바랍니다.

고생 많았습니다 오이
